src/weapon/w_030.c

```diff
@@ -66,7 +68,123 @@ INCLUDE_ASM("weapon/nonmatchings/w_030", func_ptr_80170004);
 INCLUDE_ASM("weapon/nonmatchings/w_030", func_ptr_80170008);
 
-INCLUDE_ASM("weapon/nonmatchings/w_030", func_ptr_8017000C);
+void func_ptr_8017000C(Entity* self) {
+    s16 result;
+    s16 var_v0;
+    s16 maskedParams;
+    s32 temp_a2;
+    s16 t0;
+    s32 temp_lo;
+    s32 var_v0_2;
+
+    maskedParams = (self->params >> 8) & 0x7F;
+    switch (self->step) {
+    case 0:
+        self->palette = 0x110;
+        self->unk5A = 0x64;
+        if (g_HandId != 0) {
+            self->palette += 0x18;
+            self->unk5A += 2;
+        }
+        self->flags = FLAG_UNK_08000000;
+        self->animSet = 0xA;
+        self->animCurFrame = 0x64;
+        self->zPriority = PLAYER.zPriority + 2;
+        self->facingLeft = self->ext.weapon.parent->facingLeft;
+        if (maskedParams == 1) {
+            self->posY.i.hi -= 4;
+        }
+        if (maskedParams == 0) {
+            self->posY.i.hi -= 0x10;
+        }
+        if (maskedParams == 2) {
+            self->posY.i.hi -= 28 + (rand() & 15);
+            self->posX.i.hi += D_D6000_8017A66C[D_D6000_8017CC44 % 3];
+            self->posX.i.hi += -4 + (rand() & 7);
+            self->ext.weapon.unk82 = (self->facingLeft + 1) & 1;
+            self->facingLeft = 0;
+            self->drawFlags = DRAW_COLORS;
+            self->step = 2;
+            g_api.PlaySfx(0x6BA);
+            self->ext.weapon.unk80 = rand();
+            self->ext.weapon.lifetime = ((u8)self->params * 4) + 0x30;
+            D_D6000_8017CC44++;
+        } else {
+            SetSpeedX(FIX(-5.5));
+            self->posX.val += self->velocityX * 5;
+        }
+        self->attack = 4;
+        self->attackElement = ELEMENT_HIT;
+        self->hitboxState = 2;
+        self->nFramesInvincibility = 4;
+        self->stunFrames = 4;
+        self->hitEffect = 1;
+        self->entityRoomIndex = 0;
+        g_api.func_80118894(self);
+        self->hitboxWidth = 8;
+        self->hitboxHeight = 4;
+        self->step++;
+        break;
+    case 1:
+        if (maskedParams == 2) {
+            result = g_api.func_80118B18(
+                self, self->ext.weapon.other, self->ext.weapon.unk82);
+            if (result >= 0) {
+                temp_a2 = self->ext.weapon.unk80 & 0xFFF;
+                t0 = temp_a2;
+                var_v0 = abs(t0 - result);
+                if (self->ext.weapon.unk98 > var_v0) {
+                    self->ext.weapon.unk98 = var_v0;
+                }
+
+                if (t0 < result) {
+                    if (var_v0 < 0x800) {
+                        goto block_28;
+                    }
+                    goto block_27;
+                } else if (var_v0 < 0x800) {
+                block_27:
+                    var_v0_2 = temp_a2 - (u16)self->ext.weapon.unk98;
+                } else {
+                block_28:
+                    var_v0_2 = (u16)self->ext.weapon.unk98 + temp_a2;
+                }
+                self->ext.weapon.unk80 = var_v0_2 & 0xFFF;
+            }
```

Comment:
This section probably isn't quite how the original was written, but better matching than perfect. I expect we'll come back and do a cleanup pass once the rest of this file is complete.

Reply:
Here's a cleaned up version of that section. I copied it out of the same function in Weapon 41. https://decomp.me/scratch/Q3urM